index.tsx

```diff
@@ -0,0 +1,137 @@
+import {memo, useEffect, useRef, useState} from "react";
+
+enum TimeStatus {
+    Minute,
+    Second,
+    MilliSecond
+}
+
+type TimerInfo = {
+    duration:number;
+    time:TimeStatus;
+}
+
+export const CodeReview = memo(()=>{
+    const [stopWatchTime, setStopWatchTime] = useState<number>(0)
+    const [isStopWatchRunning, setIsStopWatchRunning] = useState<boolean>(false)
+    const [timerTime, setTimerTime] = useState<number>(0)
+    const [isTimerRunning, setIsTimerRunning] = useState<boolean>(false)
+    const [timerInfo, setTimerInfo] = useState<TimerInfo>({
+        duration:0,
+        time:TimeStatus.Minute
+    })
+    const stopWatchRef = useRef<ReturnType<typeof setInterval> | undefined>(undefined);
+    const timerRef = useRef<ReturnType<typeof setInterval> | undefined>(undefined);
+
+
+    useEffect(() => {
+        if(timerTime >= 0) return;
+
+        stopTimer() // 타이머 종료
+        setTimerTime(0) // 시간 초기화
+        alert('타이머가 종료되었습니다.')
+    }, [timerTime]);
+
+
+    // 스탑워치 실행 함수
+    const startStopWatch = ()=>{
+        if(isStopWatchRunning) return;
+
+        setIsStopWatchRunning(true)
+        stopWatchRef.current = setInterval(()=>{setStopWatchTime(prev=> prev + 1)},10)
+    }
+
+    // 스탑워치 정지 함수
+    const stopStopWatch = () => {
+        if(!isStopWatchRunning) return;
+
+        setIsStopWatchRunning(false)
+        clearInterval(stopWatchRef.current)
+    }
+
+    // 타이머 정지 함수
+    const stopTimer = ()=>{
+        if(!isTimerRunning) return;
+
+        setIsTimerRunning(false)
+        clearInterval(timerRef.current)
+    }
+
+    // 타이머 실행 함수
+    const startTimer = ()=>{
+        if(isTimerRunning) return
+
+        setIsTimerRunning(true)
+        const {duration, time} = timerInfo;
+
+        if(time === TimeStatus.MilliSecond) setTimerTime(duration/10)
+        if(time === TimeStatus.Second) setTimerTime(duration * 100)
+        if(time === TimeStatus.Minute) setTimerTime(duration * 6000)
+
+        timerRef.current = setInterval(()=>{
+            setTimerTime(prev => prev - 1)
+        },10)
+    }
+
+    // 시간 포맷팅 함수
+    const getTimeFormat =(duration:number):string=>{
+        const milliseconds = String(duration % 100).padStart(2, '0');
+        const seconds = String(Math.floor((duration / 100) % 60)).padStart(2, '0');
+        const minutes = String(Math.floor(duration / (100 * 60))).padStart(2, '0');
+
+        return `${minutes} : ${seconds} : ${milliseconds}`
+    }
+
+    return (
+        <>
+            <div>
+                <h1>스탑워치</h1>
+                <div>{getTimeFormat(stopWatchTime)}</div>
+                <div >
+                    <button onClick={()=>startStopWatch()}>START</button>
+                    <button onClick={()=>stopStopWatch()}>STOP</button>
+                    <button onClick={()=>setStopWatchTime(0)}>RESET</button>
```

Comment:
```suggestion
                    <button onClick={startStopWatch}>START</button>
                    <button onClick={stopStopWatch}>STOP</button>
                    <button onClick={resetStopWatch}>RESET</button>
```
리액트에서 쓰이는 노드는 렌더링이 될 때 "참조동일성"이 무척 중요합니다.
(노드와 컴포넌트는 다릅니다!)

렌더링이 될 때 diff 알고리즘을 통해서 달라진 부분에 대해서만 렌더링을 하는데,
달라진 부분을 props가 참조하는 값을 기준으로 판단하게 됩니다.

지금처럼 익명함수로 함수를 전달하게 되면, 항상 새로운 함수를 만들어서 전달하게 되고, 무조건 리렌더링을 하게 되는거죠.

근데 아마 지금 제가 위에서 제안한 것 처럼 바꾼다고 해도 컴포넌트가 실행될 때 다시 함수가 재선언 되기 때문에 어차피 새로운 함수를 만들게 됩니다.

이 때 새로운 함수를 만들지 않는 방법은 무엇이 있을까요?
한 번 고민해보세요!